feat: [M3-7284, M3-7361] - Create VPC drawer in Linode Create flow; Update widths for Create VPC and Create Firewall buttons

coliu-akamai opened the pull request:
## Description 📝
- Add Create VPC drawer to the Linode Create flow. Now, if a user wants to create a VPC while creating a linode, they can do so while staying on the same page. No more new tabs!
- Update 11/6: including M3-7361 in this ticket (as it'd been caught and handled here earlier anyway)

## Changes  🔄
List any change relevant to the reviewer.
- Add `CreateVPCDrawer`
- Abstracted a lot of common logic between `VPCCreate` page and `VPCCreateDrawer`, creating a new custom hook and other components
- Added some tests for new components
- Created new constants in the VPC's constants file for some text 
- some VPC Create page cosmetic changes - Right aligned 'Create VPC' button on VPC create page, removed divider between last subnet and 'Add another subnet' button

## Preview 📷
**Include a screenshot or screen recording of the change**
(the grey shadow in the second img is from my computer's menu bar)

Region that doesn't support vpcs selected:
![image](https://github.com/linode/manager/assets/139280159/503c1724-68f2-44fa-89d9-f676ab9c9690)


| drawer img 1  | drawer img 2   |
| ------- | ------- |
| ![image](https://github.com/linode/manager/assets/139280159/8af9f2d6-5dad-4893-9929-f58a64ff3c06) | ![image](https://github.com/linode/manager/assets/139280159/559a370b-72c9-4b32-bea9-1b208f0244f5) |

## How to test 🧪

### Prerequisites
- Using the dev environment, with vpc admin tags

### Verification steps 

On the Linode Create page
- Verify that if you do not have a region selected, or have a [region that does not support VPCs selected](https://github.com/linode/manager/pull/9787), the Create VPC does not appear and there's instead text that says "VPC is not available in the selected region."
- If the Create VPC link shows up:
  - Verify that the region you've chosen is pre-selected and the region drop down is disabled
  - Verify VPC Create functionality (it should work the same as the VPC Create page EXCEPT that you must have at least 1 subnet in this flow)
  - Verify that after creating a VPC, the VPC is selected in the VPC panel and that creating a linode with that VPC works as expected

On the VPC Create page
- verify the 'Create VPC' button is now right aligned, not left aligned
- verify there's no divider between the last subnet and the 'Add another subnet' button (and that there's increased spacing between the last subnet and add subnet button)
- verify that besides those changes ^, the VPC Create page looks the same
- verify that the VPC Create page still functions the same 

On the VPC details page
- verify the Create Subnet drawer still looks the same (and still functions the same, although its functionality shouldn't have been affected by these changes)

### Tests
- `yarn test VPC` to run all the VPC related tests
- `yarn cy:run -s 'cypress/e2e/core/vpc/vpc-create.spec.ts`  to make sure there are no regressions here

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [x] 📑 Providing or updating our documentation
- [x] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [x] ♿  Providing accessibility support

coliu-akamai commented on packages/manager/src/hooks/useCreateVPC.ts:
```diff
@@ -0,0 +1,223 @@
+  handleSelectVPC?: (vpcId: number) => void;
+  onDrawerClose?: () => void;
+  pushToVPCPage?: boolean;
+  selectedRegion?: string;
```
created custom hook for shared form logic between `VPCCreate` and `VPCCreateDrawer`. There are some slight differences in the functionality of the drawer and page though:
- the page will push to the newly created vpc's detail page, but the drawer only makes the newly created vpc the selected vpc for linode-create flow and closes the drawer
- the drawer's region is pre-selected, based on the selected linode region

The main changes from the old create vpc logic are at lines 133-139 and line 185.

dwiley-akamai replied:
I might be missing something -- isn't L185 here just `description: '',` ?

coliu-akamai replied:
Ah I might have made some changes to this file that changed the lines - the main difference is for the region initial value, it's now `region: selectedRegion ?? '',` instead of `region: '',`

coliu-akamai commented on packages/manager/src/features/VPCs/VPCCreate/FormComponents/CannotCreateVPCNotice.test.tsx:
```diff
@@ -0,0 +1,15 @@
+import * as React from 'react';
```
Similar to the new custom hook, consolidated shared components for the form itself into this new folder. (added tests for them as well altho they're pretty repetitive since SubnetNode, MultipleSubnetInput, and VPCCreate already tested how they render).

However, because there are some styling differences between the `VPCCreateDrawer` and `VPCCreate` page, I did end up having to pass an `isDrawer` prop down ;-; (renamed specifically to `isVPCCreateDrawer` for the SubnetNode component, since the Create Subnet drawer also used a `SubnetNode`, not just the `VPCCreateDrawer`)

coliu-akamai commented on packages/manager/src/features/VPCs/VPCCreate/SubnetNode.tsx:
```diff
@@ -17,6 +18,7 @@ interface Props {
   // extra props enable SubnetNode to be an independent component or be part of MultipleSubnetInput
   // potential refactor - isRemoveable, and subnetIdx & remove in onChange prop
   idx?: number;
+  isCreateVPCDrawer?: boolean;
```
See comment on `CannotCreateVPCNotice` for reasoning behind this prop

coliu-akamai commented on packages/manager/src/features/VPCs/VPCCreate/SubnetNode.tsx:
```diff
@@ -52,49 +61,55 @@ export const SubnetNode = (props: Props) => {
     onChange(subnet, idx, isRemovable);
   };
 
+  const showRemoveButton = isCreateVPCDrawer
+    ? idx !== 0 && isRemovable
+    : isRemovable;
```
For the VPCCreateDrawer, we do not want the remove button to show for the first subnet

coliu-akamai reviewed (commented):
Since we disable the Create VPC link in the vpc panel if the linode's region isn't selected or if the region selected doesn't support VPCs, sometimes the link just looks like text (below). Speaking with Andrew about potential friendlier UX options!

![image](https://github.com/linode/manager/assets/139280159/a611beae-0d46-4e47-b8e2-8866f1d8b0e4)

hana-linode commented:
Is there a reason for keeping both drawer and page versions? For example, Create Firewalls is just a drawer

coliu-akamai commented:
hm...iirc I think people still liked having the VPC create page? I think there was a thread about it somewhere

dwiley-akamai commented:
Going to review this first thing tomorrow!

> Is there a reason for keeping both drawer and page versions? For example, Create Firewalls is just a drawer

We've generally been moving away from drawer entity creation for a couple of years at least -- before this, firewalls were the only exception. This happens to be a specific use case where having a drawer makes sense so I don't think we'd be moving in the direction of getting rid of the Create VPC page entirely

dwiley-akamai reviewed (commented):
The basic functionality is looking good so far. I noticed a couple of UI issues:

1) In the drawer, when you click into the "Description" field and start typing, the cursor and text appear above the field, overlaid on the field label:

<details>
<summary>Video</summary>

I recorded the video before I typed anything into the field, so that part isn't shown

https://github.com/linode/manager/assets/114682940/04c23ef3-66c5-41d2-9a93-0e1975246bf6

</details>

2. The "Create VPC" link takes up the full width of the panel almost, so clicking to the far right of it still opens up the drawer

![Screenshot 2023-10-31 at 12 46 48 PM](https://github.com/linode/manager/assets/114682940/236ecfc0-77ba-43a9-aa58-24004a2fea67)

3. On the Linode Create page, before you select a region, scrolling down to the VPC panel has `VPC is not available in the selected region.` text (we should handle for the case where no region has been selected yet with specific text)

coliu-akamai commented:
👀  Thanks @dwiley-akamai

1. For the description field, I removed the specified height applied to the textArea and the cursor is now where it should be. However, this makes the textfield area larger than the UX mock -- I was looking into [props like `maxRows` and `rows`](https://mui.com/material-ui/react-text-field/#multiline) that the underlying mui component takes, but strangely the behavior I'm seeing is not matching the mui example. Will try to keep on looking into this to see if we can get the correct height for this drawer but am unsure
3. Added placeholder text for now, will check with Andrew for the preferred copy

coliu-akamai commented:
@dwiley-akamai spoke with Andrew -- we'll be hiding the text/Create drawer link if no region is selected. The VPC panel will look like this in that case:
![image](https://github.com/linode/manager/assets/139280159/04c7535a-3083-4d26-a0ba-98dbde32d50e)

1 - Still having trouble trying to get the description field's height match the mockup. Continuing to look into it 😅 
2 - 'Create VPC' link should no longer take up the full width of the panel. I saw that the Create Firewall link had similar behavior -- both should now be fixed 🎉

coliu-akamai commented on packages/manager/cypress/e2e/core/databases/update-database.spec.ts:
Thanks Dajahi! Gonna opt to leave it, and once it gets merged in hopefully it shouldn't reappear in someone else's pr 😅

dwiley-akamai reviewed (commented):
Still need to finish combing through the code, but:

All tests pass ✅
Linode Create page ✅
- No region selected --> "Create VPC" does not appear ✅
- Region selected where VPC is not enabled --> text displayed that says "VPC is not available in the selected region." ✅
- "Create VPC" link shows up with chosen region pre-selected, Region dropdown is disabled ✅
  - VPC Create drawer functionality ✅
- VPC created via drawer, newly-created VPC is selected in the VPC panel, Linode creation works as expected ✅
 
VPC Create page ✅
- "Create VPC" button is right-aligned ✅
- No divider between the last subnet and the "Add Another Subnet" button (and that there's increased spacing between the last subnet and add subnet button) ✅

VPC Details --> Create Subnet drawer ✅

dwiley-akamai commented on packages/manager/src/features/Linodes/LinodesCreate/VPCPanel.tsx:
```diff
@@ -157,165 +162,185 @@ export const VPCPanel = (props: VPCPanelProps) => {
+          )}
+        </Stack>
+      </Paper>
+      {isVPCCreateDrawerOpen &&
+        from === 'linodeCreate' &&
+        regionSupportsVPCs && (
```
Is there a reason just checking for `isVPCCreateDrawerOpen` wouldn't be enough here? We don't display the "Create VPC" link if a region isn't selected or the selected region doesn't support VPCs.

coliu-akamai replied:
Hm, I think I might have added them as a way to be extra sure but you're right -- will remove the unnecessary logic!

dwiley-akamai commented on packages/manager/src/features/VPCs/VPCCreate/SubnetNode.tsx:
```diff
@@ -52,49 +61,55 @@ export const SubnetNode = (props: Props) => {
+        {showRemoveButton && (
           <Grid xs={1}>
-            <StyledButton onClick={removeSubnet} aria-label="Remove Subnet">
+            <StyledButton aria-label={`Remove Subnet`} onClick={removeSubnet}>
```
Did the linter change this line?

```suggestion
            <StyledButton aria-label="Remove Subnet" onClick={removeSubnet}>
```

coliu-akamai replied:
Think so!

dwiley-akamai commented on packages/manager/src/hooks/useCreateVPC.ts:
```diff
@@ -0,0 +1,225 @@
+
+    for (let i = 0; i < formik.values.subnets.length; i++) {
+      const { ip, label } = formik.values.subnets[i];
+      // if we are inside the VPCCreateDrawer, we force the first subnet to
+      // always be included in the payload, even if its fields are empty.
```
Is that for validation purposes since subnets are required in the drawer? If so, we should clarify that in the comment

dwiley-akamai reviewed (commented):
Confirmed the UX issues I mentioned last time have been resolved ✅

coliu-akamai commented on packages/manager/src/features/VPCs/VPCCreate/FormComponents/VPCSpecificContent.tsx:
```diff
@@ -0,0 +1,63 @@
+  values: CreateVPCFieldState;
+}
+
+export const VPCSpecificContent = (props: Props) => {
```
Good point -- maybe `VPCMainContent`? But I feel like that has similar issues to `VPCSpecificContent` 😅 
I'll sleep on it haha

coliu-akamai replied:
Here are some other potential names:
```
<VPCInitialContent>
<VPCTopContent>
<VPCFirstContent>
<VPCPrimaryContent>
```
They're all pretty similar - I think VPCPrimaryContent could have the same issue as the original name. VPCInitialContent and VPCTopContent could work - I prefer VPCInitialContent a bit more. What are your thoughts / do you have other name ideas? 👀

dwiley-akamai replied:
Maybe `VPCTopSectionContent`? Trying not to overthink this or get too verbose haha. At first I was thinking `VPCInitialContent` was the way to go (it moves us in the direction of describing the fields' position within the page/form) but "initial" can have some connotations related to the lifecycle of fields/pages and the pre-filling of data, etc. So I'd vote for something like `VPCTopSectionContent`.

coliu-akamai replied:
VPCTopSectionContent sounds good! 🎉

hana-linode reviewed (commented):
Functionality LGTM. Since the Create VPC flow is now on the same page, I don't think we need the alwaysrefetch logic anymore 
https://github.com/linode/manager/blob/develop/packages/manager/src/queries/vpcs.ts#L35

https://github.com/linode/manager/assets/115299789/50f9b4e2-1006-45df-a617-47786113c33b

coliu-akamai commented:
@hana-linode removed the refetch logic, thanks for catching!

hana-linode reviewed (approved):
On the Linode Create page
- ✅ Verify that if you do not have a region selected, or have a [region that does not support VPCs selected](https://github.com/linode/manager/pull/9787), the Create VPC does not appear and there's instead text that says "VPC is not available in the selected region."
- If the Create VPC link shows up:
  - ✅ Verify that the region you've chosen is pre-selected and the region drop down is disabled
  - ✅ Verify VPC Create functionality (it should work the same as the VPC Create page EXCEPT that you must have at least 1 subnet in this flow)
  - ✅ Verify that after creating a VPC, the VPC is selected in the VPC panel and that creating a linode with that VPC works as expected
- ✅ VPCs are no longer refetched on window refocus

On the VPC Create page
- ✅ verify the 'Create VPC' button is now right aligned, not left aligned
- ✅ verify there's no divider between the last subnet and the 'Add another subnet' button (and that there's increased spacing between the last subnet and add subnet button)
- ✅ verify that besides those changes ^, the VPC Create page looks the same
- ✅ verify that the VPC Create page still functions the same 

On the VPC details page
- ✅ verify the Create Subnet drawer still looks the same (and still functions the same, although its functionality shouldn't have been affected by these changes)
